main.go

```diff
@@ -442,7 +442,27 @@ func setupReconcilers(ctx context.Context, mgr ctrl.Manager) webhooks.ClusterCac
 		},
 	})
 	if err != nil {
-		setupLog.Error(err, "unable to create unstructured caching client")
+		setupLog.Error(err, "Unable to create unstructured caching client")
+		os.Exit(1)
+	}
+
+	// Setup a separate cache for the metadata watches to secrets.
+	// This way the watch does not use the LabelSelector defined at the cache which
+	// would filter to secrets having the cluster label, because secrets referred
+	// by ClusterResourceSet or ExtensionConfig are not specific to a single cluster.
+	partialSecretCache, err := cache.New(mgr.GetConfig(), cache.Options{
```

Comment:
q: should this be `allSecretCache` instead of `partialSecretCache` (nothing in the definition points to partial)
q: is there a way to make sure this cache is used only for Secrets (I think not, but might be we can enforce this with a DefaultTransformerFunc that always returns error)
q: should we use TransformStripManagedFields for secrets? (not necessary, but it doesn't hurt)

cc @sbueringer

Reply:
the intention why I named it `partialSecretCache` is that we tend to only use it for `PartialObjectMetadata` watches/objects. Maybe I should add that information to the comment?

Reply:
I like the idea of adding a DefaultTransformerFunc and implemented it.

This way we can make sure to not mis-use the cache 👍

Reply:
Does PartialObjectMeta contain a relevant amount of managed fields?

Reply:
What happens if we hit the error cases now? (does the controller fail? are we getting not founds on get? anything else?)

Is the behavior good enough to make sure we never use the cache for the wrong purpose? Or will it just mean that some of our code doesn't work.

Reply:
I'm wondering if it would be better if we just panic if someone tries to use this cache for the wrong gvk

Reply:
It's currently only logging but retrying.

```
E0604 05:56:24.840511      17 reflector.go:150] pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/reflector.go:232: Failed to watch *v1.PartialObjectMetadata: unable to sync list result: couldn't enqueue object: cache expected to only get Secrets, got &TypeMeta{Kind:ConfigMap,APIVersion:v1,}
```

I'll adjust to do a panic instead.

Reply:
Now it observes the panic, prints the trace and gets stuck (which I think is better than just logging the error)

```
E0604 05:58:32.951254      62 runtime.go:79] Observed a panic: &errors.errorString{s:"cache expected to only get Secrets, got &TypeMeta{Kind:ConfigMap,APIVersion:v1,}"} (cache expected to only get Secrets, got &TypeMeta{Kind:ConfigMap,APIVersion:v1,})
goroutine 313 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x271ef40, 0x40003a6bf0})
	/Users/schlotterc/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/runtime/runtime.go:75 +0xdc
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x0})
	/Users/schlotterc/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/runtime/runtime.go:49 +0xb4
panic({0x271ef40?, 0x40003a6bf0?})
	/Users/schlotterc/.bin/go-archive/go1.22.1.darwin-arm64/src/runtime/panic.go:770 +0xf0
main.setupReconcilers.func1({0x2aaede0, 0x40008f37a0})
	/Users/schlotterc/go/src/sigs.k8s.io/cluster-api/main.go:455 +0x2f0
k8s.io/client-go/tools/cache.(*DeltaFIFO).queueActionLocked(0x4000362f20, {0x2aeb978, 0x8}, {0x2aaede0, 0x40008f37a0})
	/Users/schlotterc/go/pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/delta_fifo.go:456 +0x15c
k8s.io/client-go/tools/cache.(*DeltaFIFO).Replace(0x4000362f20, {0x4000292280, 0x13, 0x13}, {0x40007e99b8, 0x4})
	/Users/schlotterc/go/pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/delta_fifo.go:641 +0x390
k8s.io/client-go/tools/cache.(*Reflector).syncWith(0x4000988a80, {0x4000292140, 0x13, 0x13}, {0x40007e99b8, 0x4})
	/Users/schlotterc/go/pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/reflector.go:706 +0x1d0
k8s.io/client-go/tools/cache.(*Reflector).list(0x4000988a80, 0x4000a1ae40)
	/Users/schlotterc/go/pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/reflector.go:577 +0xe68
k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0x4000988a80, 0x4000a1ae40)
	/Users/schlotterc/go/pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/reflector.go:353 +0x344
k8s.io/client-go/tools/cache.(*Reflector).Run.func1()
	/Users/schlotterc/go/pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/reflector.go:298 +0x30
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x40009ede38)
	/Users/schlotterc/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:226 +0x48
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x4000a59e38, {0x2d710c0, 0x4000290b90}, 0x1, 0x4000a1ae40)
	/Users/schlotterc/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/backoff.go:227 +0xa0
k8s.io/client-go/tools/cache.(*Reflector).Run(0x4000988a80, 0x4000a1ae40)
	/Users/schlotterc/go/pkg/mod/k8s.io/client-go@v0.30.1/tools/cache/reflector.go:297 +0x24c
k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1()
	/Users/schlotterc/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/wait.go:55 +0x34
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
	/Users/schlotterc/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/wait.go:72 +0xa8
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 305
	/Users/schlotterc/go/pkg/mod/k8s.io/apimachinery@v0.30.1/pkg/util/wait/wait.go:70 +0xc4
2024-06-04T05:58:32Z error layer=rpc writing response:write tcp [::1]:30000->[::1]:46760: use of closed network connection
█
```